integrations/webhook/src/index.ts

```diff
@@ -0,0 +1,153 @@
+import { createIntegration, Logger } from '@gitbook/runtime';
+import { Event } from '@gitbook/api';
+
+import { configComponent } from './config';
+import {
+    MAX_RETRIES,
+    REQUEST_TIMEOUT,
+    WebhookRuntimeContext,
+    retryWithDelay,
+    EventType,
+    EVENT_TYPES,
+} from './common';
+
+const logger = Logger('webhook');
+
+/**
+ * Common webhook delivery handler for all event types
+ */
+const handleWebhookEvent = async (event: Event, context: WebhookRuntimeContext) => {
+    const { environment } = context;
+    const spaceInstallation = environment.spaceInstallation;
+
+    if (!spaceInstallation) {
+        logger.debug('No space installation found');
+        return;
+    }
+
+    const config = spaceInstallation.configuration;
+
+    // Check if this event type is supported and enabled
+    if (!EVENT_TYPES.includes(event.type as EventType) || !config[event.type as EventType]) {
+        logger.debug(`Event ${event.type} is not enabled`);
+        return;
+    }
+
+    // Prepare webhook payload
+    const jsonPayload = JSON.stringify(event);
+
+    // Add HMAC signature for webhook verification
+    const encoder = new TextEncoder();
+    const key = await crypto.subtle.importKey(
+        'raw',
+        encoder.encode(config.secret),
+        { name: 'HMAC', hash: 'SHA-256' },
+        false,
+        ['sign'],
+    );
+    const signatureBuffer = await crypto.subtle.sign('HMAC', key, encoder.encode(jsonPayload));
+    const signature = Array.from(new Uint8Array(signatureBuffer))
+        .map((b) => b.toString(16).padStart(2, '0'))
+        .join('');
+
+    const sendWebhookWithRetry = async (retryCount = 0): Promise<void> => {
+        const startTime = Date.now();
+        try {
+            const response = await fetch(config.webhookUrl, {
+                method: 'POST',
+                headers: {
+                    'Content-Type': 'application/json',
+                    'User-Agent': 'GitBook-Webhook',
+                    'X-GitBook-Signature': `sha256=${signature}`,
+                },
+                body: jsonPayload,
+                signal: AbortSignal.timeout(REQUEST_TIMEOUT),
+            });
+
+            if (!response.ok) {
+                const errorMessage = `Webhook delivery failed: ${response.status} ${response.statusText}`;
+                logger.error(errorMessage, {
+                    url: config.webhookUrl,
+                    eventType: event.type,
+                    retryCount,
+                    status: response.status,
+                    statusText: response.statusText,
+                });
+
+                // Retry on server errors (5xx) and rate limiting (429)
+                if (
+                    retryCount < MAX_RETRIES &&
+                    (response.status >= 500 || response.status === 429)
+                ) {
+                    return retryWithDelay(
```

Comment:
How long could the integration script be delayed in the worst-case retry scenario? I’m not sure what the upper limit is for Cloudflare Workers.

Not sure if it's worth but we could consider using the `api.integrations.queueIntegrationTask` API (similar to how the GitHub integration works) to handle this.

But if the whole retry window is reasonable then probably no need.

Reply:
Good question!

So for the exponential backoff, counting also the worst possible jitter, it should be:
1000+100 + 2000+200 + 4000+400 = 7700 ms

But, if there's a weird connectivity issue we could hit the request timeouts and then it becomes:
10000+1100 + 10000+2200 + 10000+4400  for a total of almost 38 seconds 

I think workers should be totally fine with these waits

Reply:
I'm going to look into `api.integrations.queueIntegrationTask` and see how that would work

Reply:
I implemented `queueIntegrationTask` for retries and added a bunch of tests. I'm going to build a stupid server to test it properly tomorrow manually as well